django/contrib/postgres/search.py

```diff
@@ -0,0 +1,156 @@
+            raise TypeError('SearchVector can only be combined with other SearchVectors')
+        combined = super(SearchVector, self)._combine(other, connector, reversed, node)
+        combined.config = self.config
+        return combined
+
+
+class SearchQuery(Value):
+    config = None
+    invert = False
+    _output_field = SearchQueryField()
+
+    BITAND = '&&'
+    BITOR = '||'
+
+    def __init__(self, value, output_field=None, **extra):
+        self.config = extra.pop('config', self.config)
+        self.invert = extra.pop('invert', self.invert)
+        super(SearchQuery, self).__init__(value, output_field=output_field)
+
+    def as_sql(self, compiler, connection):
+        params = [self.value]
+        if self.config:
+            template = "plainto_tsquery(%s, %s)"
+            params = [self.config, self.value]
+        else:
+            template = "plainto_tsquery(%s)"
+        if self.invert:
+            template = "!!({})".format(template)
+        return template, params
+
+    def _combine(self, other, connector, reversed, node=None):
+        combined = super(SearchQuery, self)._combine(other, connector, reversed, node)
+        combined.output_field = SearchQueryField()
+        return combined
+
+    # On Combinable, these are not implemented to reduce confusion with Q. In
+    # this case we are actually (ab)using them to do logical combination so
+    # it's consistent with other usage in Django.
+    def __or__(self, other):
+        return self._combine(other, self.BITOR, False)
+
+    def __ror__(self, other):
+        return self._combine(other, self.BITOR, True)
+
+    def __and__(self, other):
+        return self._combine(other, self.BITAND, False)
+
+    def __rand__(self, other):
+        return self._combine(other, self.BITAND, True)
+
+    def __invert__(self):
+        extra = {
+            'invert': not self.invert,
+            'config': self.config,
+        }
+        obj = type(self)(self.value, **extra)
+        return obj
+
+
+class SearchRank(Func):
+    function = 'ts_rank'
+    _output_field = FloatField()
+
+    def __init__(self, vector, query, **extra):
+        if not hasattr(vector, 'resolve_expression'):
+            vector = SearchVector(vector)
+        if not hasattr(query, 'resolve_expression'):
+            query = SearchQuery(query)
+        weights = extra.get('weights', None)
+        if weights is not None and not hasattr(weights, 'resolve_expression'):
+            weights = Value(weights)
+        self.weights = weights
+        super(SearchRank, self).__init__(vector, query, **extra)
+
+    def as_sql(self, compiler, connection, function=None, template=None):
+        extra_params = []
+        if template is None and self.extra.get('weights'):
+            if self.weights:
+                template = '%(function)s(%(weights)s, %(expressions)s)'
+                weight_sql, extra_params = compiler.compile(self.weights)
+                self.extra['weights'] = weight_sql
```

Comment:
Probably want to use `**extra_context`  added in 53361589905c7c07692cd77f398ce6cb5ac39779 rather than mutate `self`.